File: server_integration_tests/clients/kotlin_client/client-app/src/test/kotlin/dev/graphlink/server/HoistArgsTest.kt
package dev.graphlink.server

import dev.graphlink.server.generated.client.GraphLinkClient
import dev.graphlink.server.generated.inputs.CatalogFieldArgs1
import dev.graphlink.server.generated.inputs.FeedFieldArgs
import dev.graphlink.server.generated.inputs.SearchFieldArgs
import dev.graphlink.server.generated.inputs.StoreFieldArgs
import kotlinx.coroutines.test.runTest
import kotlin.test.BeforeTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class HoistArgsTest {

    private lateinit var client: GraphLinkClient

    @BeforeTest
    fun setUp() {
        client = Fixtures.newClient()
    }

    @Test
    fun manyRequiredHoistedArgsInputRenamedCatalogFieldArgs1OnCollision() = runTest {
        val catalog = client.queries.catalog(
            CatalogFieldArgs1(productsCategory = "books", productsLimit = 3, productsOffset = 0),
        ).catalog
        assertEquals(3, catalog.products.size)
        assertTrue(catalog.products.all { it.name.startsWith("books") })
    }

    @Test
    fun allOptionalHoistedArgsCanBeOmittedDefaultItemsLimit10() = runTest {
        val feed = client.queries.feed().feed
        assertEquals(10, feed.items.size)
    }

    @Test
    fun optionalHoistedArgsCanBeSupplied() = runTest {
        val feed = client.queries.feed(FeedFieldArgs(itemsLimit = 4, itemsSort = "z")).feed
        assertEquals(4, feed.items.size)
        assertTrue(feed.items.first().name.startsWith("z"))
    }

    @Test
    fun nestedDepthHoistingCollapsesIntoOneInput() = runTest {
        val store = client.queries.store(StoreFieldArgs(shelvesFloor = 2, productsLimit = 3)).store
        assertEquals(2, store.shelves.size)
        assertEquals(3, store.shelves.first().products.size)
    }

    @Test
    fun paramNameCollisionUsesFieldArgs1ForSyntheticArg() = runTest {
        val search = client.queries.search("q", SearchFieldArgs(hitsLimit = 3)).search
        assertEquals("q", search.id)
        assertEquals(3, search.hits.size)
    }
}

File: server_integration_tests/clients/swift_client/Tests/GraphLinkGeneratedTests/HoistArgsTest.swift
import XCTest
@testable import GraphLinkGenerated

final class HoistArgsTest: XCTestCase {
    var client: GraphLinkClient!

    override func setUp() {
        client = Fixtures.newClient()
    }

    func testManyRequiredHoistedArgsInputRenamedCatalogFieldArgs1OnCollision() async throws {
        let catalog = try await client.queries.catalog(
            fieldArgs: CatalogFieldArgs1(productsCategory: "books", productsLimit: 3, productsOffset: 0)
        ).catalog
        XCTAssertEqual(catalog.products.count, 3)
        XCTAssertTrue(catalog.products.allSatisfy { $0.name.hasPrefix("books") })
    }

    func testAllOptionalHoistedArgsCanBeOmittedDefaultItemsLimit10() async throws {
        let feed = try await client.queries.feed().feed
        XCTAssertEqual(feed.items.count, 10)
    }

    func testOptionalHoistedArgsCanBeSupplied() async throws {
        let feed = try await client.queries.feed(fieldArgs: FeedFieldArgs(itemsLimit: 4, itemsSort: "z")).feed
        XCTAssertEqual(feed.items.count, 4)
        XCTAssertTrue(feed.items.first!.name.hasPrefix("z"))
    }

    func testNestedDepthHoistingCollapsesIntoOneInput() async throws {
        let store = try await client.queries.store(fieldArgs: StoreFieldArgs(shelvesFloor: 2, productsLimit: 3)).store
        XCTAssertEqual(store.shelves.count, 2)
        XCTAssertEqual(store.shelves.first!.products.count, 3)
    }

    func testParamNameCollisionUsesFieldArgs1ForSyntheticArg() async throws {
        let search = try await client.queries.search(fieldArgs: "q", fieldArgs1: SearchFieldArgs(hitsLimit: 3)).search
        XCTAssertEqual(search.id, "q")
        XCTAssertEqual(search.hits.count, 3)
    }
}
